api/repositories/service_offering_repository.go

```diff
@@ -117,3 +157,112 @@ func offeringToRecord(offering korifiv1alpha1.CFServiceOffering) ServiceOffering
+	if err != nil {
+		return fmt.Errorf("failed to list service instances: %w", err)
+	}
+
+	for _, instance := range serviceInstances {
+		err = k8s.PatchResource(ctx, userClient, &instance, func() {
+			controllerutil.RemoveFinalizer(&instance, korifiv1alpha1.CFManagedServiceInstanceFinalizerName)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to remove finalizer for service instance: %s, %w", instance.Name, apierrors.FromK8sError(err, ServiceInstanceResourceType))
+		}
+
+		if err = userClient.Delete(ctx, &instance); err != nil {
+			return fmt.Errorf("failed to delete service instance: %w", apierrors.FromK8sError(err, ServiceInstanceResourceType))
+		}
+
+	}
+
+	serviceBindings, err := r.fetchServiceBindings(ctx, userClient, authorizedSpaceNamespacesIter, planGUIDs)
+	if err != nil {
+		return fmt.Errorf("failed to list service bindings: %w", err)
+	}
+
+	for _, binding := range serviceBindings {
+		err = k8s.PatchResource(ctx, userClient, &binding, func() {
+			controllerutil.RemoveFinalizer(&binding, korifiv1alpha1.CFServiceBindingFinalizerName)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to remove finalizer for service binding: %s, %w", binding.Name, apierrors.FromK8sError(err, ServiceBindingResourceType))
+		}
+	}
+
+	return nil
+}
+
+func (r *ServiceOfferingRepo) deleteServicePlans(ctx context.Context, userClient client.WithWatch, offeringGUID string) ([]string, error) {
+	var planGUIDs []string
+	plans := &korifiv1alpha1.CFServicePlanList{}
+
+	if err := userClient.List(ctx, plans, client.InNamespace(r.rootNamespace), client.MatchingLabels{
+		korifiv1alpha1.RelServiceOfferingGUIDLabel: offeringGUID,
+	}); err != nil {
+		return []string{}, fmt.Errorf("failed to list service plans: %w", err)
+	}
+
+	for _, plan := range plans.Items {
+		planGUIDs = append(planGUIDs, plan.Name)
+		if err := userClient.Delete(ctx, &plan); err != nil {
+			return []string{}, apierrors.FromK8sError(err, ServicePlanResourceType)
+		}
+	}
+
+	return planGUIDs, nil
+}
+
+func (r *ServiceOfferingRepo) fetchServiceInstances(ctx context.Context, userClient client.WithWatch, authorizedNamespaces itx.Iterator[string], planGUIDs []string) ([]korifiv1alpha1.CFServiceInstance, error) {
+	var serviceInstances []korifiv1alpha1.CFServiceInstance
+
+	for _, ns := range authorizedNamespaces.Collect() {
+		instances := new(korifiv1alpha1.CFServiceInstanceList)
+
+		err := userClient.List(ctx, instances, client.InNamespace(ns))
+		if err != nil {
+			return []korifiv1alpha1.CFServiceInstance{}, fmt.Errorf("failed to list service instances: %w", err)
+		}
+
+		filtered := itx.FromSlice(instances.Items).Filter(func(serviceInstance korifiv1alpha1.CFServiceInstance) bool {
+			return tools.EmptyOrContains(planGUIDs, serviceInstance.Spec.PlanGUID)
+		}).Collect()
+
+		serviceInstances = append(serviceInstances, filtered...)
+	}
+
+	return serviceInstances, nil
+}
```

Comment:
We can combine the fetching of the service instances and bindings under one loop, so we do not need to do it twice, but this way it seems easier to read. Let me know if I should change it.

Reply:
Agree, two methods is more readable